### ISSUE TITLE
Symfony\Component\Security\Http\Firewall\ExceptionListener overrides AccessDeniedListener

### ISSUE THREAD

**Issue by bblue** [FriendsOfSymfony/FOSRestBundle#1538]
The logic in the `AccessDeniedListener` is effectively overriden by `Symfony\Component\Security\Http\Firewall\ExceptionListener`

The priority of the Symfony listener is lower than that of FOSRest, thus the following method in Symfony's listener overrides the `$event->setException($e);`

```
private function handleAuthenticationException(GetResponseForExceptionEvent $event, AuthenticationException $exception)
{
    if (null !== $this->logger) {
        $this->logger->info('An AuthenticationException was thrown; redirecting to authentication entry point.', array('exception' => $exception));
    }

    try {
        $event->setResponse($this->startAuthentication($event->getRequest(), $exception));
    } catch (\Exception $e) {
        $event->setException($e);
    }
}
```

This will trigger a redirect to the login page, and in some cases (my case) this will trigger yet another authentication exception (due to faulty JWT token for instance), but this new exception will be caught in the above method, the `$event->exception` will be overridden, and hence no conversion to the http exceptions are done. I will still get JSON data, but it will come with a 500 error.

**Comment by geoffreytran** [comment]
Running into this same issue here

**Comment by pppdns** [comment]
Same here, although I get HTML data (coming from [RedirectResponse#L85](https://github.com/symfony/http-foundation/blob/2c8a1e1/RedirectResponse.php#L85)))

**Comment by lsmith77** [comment]
FYI I created a PR but I am not sure yet what the side effects will be https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1692

**Comment by scott-r-lindsey** [comment]
Any movement on this?

**Comment by bblue** [comment]
Just realized I had been referenced in this thread a while ago. Haven't worked on that project for a loooong time, and its dependencies and symfony has not been updated since 2017. 

I recall initally making a compiler pass to override the FOSRest listener, but was not happy with that solution. I ended up instead solving the root cause that threw my initial exception (not linked to fosrest), thus I really just side-steped the real bug, but I then would not have to override any FOSRest classes.

So to summarize: you can probably avoid a particular use-case error with a compiler pass, but it would be fragile. Unfortunately I have deleted my compiler pass object, so I can't tell you what I actually wrote in it... I probably just rethrew the exception.

I'm guessing the PR by @lsmith77 is the way to go, but have not looked at it.

**Comment by xabbuh** [comment]
I was trying to write a test case for this issue to ensure that #2109 is indeed the right fix, but unfortunately I wasn't able to do so. Can anyone of you suffering from this issue please create a small example application that allows to do so?

**Comment by lsmith77** [comment]
i will try to have a look today
